Add multi-instance RPMsg and IPC Service

maje-emb opened the pull request:
The aim of this PR is to provide a generic API for IPC communications. 
IPC Service allow plugging in different transport backends.

In addition, a multi-instance PRMsg was created and used as a backend for the IPC Service. 
Each instance is allocated a separate piece of shared memory.
Multiple instances provide independent message processing.
Each instance has its own work_q.

Sample of using both multi-instance RPMsg and IPC service for nrf5340 have been prepared.

To build an IPC Service or multi-instance RPMsg sample for nrf5340:
```
west build -b nrf5340dk_nrf5340_cpuapp
west build -b nrf5340dk_nrf5340_cpunet
```

Relates to #20419.

maje-emb committed:
ipc: Add multiple instances RPMsg

This patch implements a service that adds multiple instances
capabilities to RPMsg.
Each instance is allocated a separate piece of shared memory.
Multiple instances provide independent message processing.
Each instance has its own work_q.

Signed-off-by: Marcin Jeliński <marcin.jelinski@nordicsemi.no>

maje-emb committed:
samples: subsys: ipc: Add a sample for using multi-instance RPMsg

Demonstating usage of multiple instance of RPMsg

Signed-off-by: Marcin Jeliński <marcin.jelinski@nordicsemi.no>

maje-emb committed:
ipc: Added IPC Service to support different transport backends

IPC Service allow plugging in different transport backends.
Specifies a generic API that is implemented by the backend.

Signed-off-by: Marcin Jeliński <marcin.jelinski@nordicsemi.no>

maje-emb committed:
ipc: ipc_service: RPMsg multi-instance as backend for IPC Service

Implementation of the backend for IPC Service.
Multi-instance RPMsg was used as a backend.

Signed-off-by: Marcin Jeliński <marcin.jelinski@nordicsemi.no>

maje-emb committed:
samples: subsys: ipc: Add a sample for using IPC Service

Demonstating usage of IPC Service. Multi-instance RPMsg
was ued as a backed for IPC Service.

Signed-off-by: Marcin Jeliński <marcin.jelinski@nordicsemi.no>

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -0,0 +1,91 @@
+ */
+struct ipc_ept;
+
+/** @brief Endpoint configuration structure */
+struct ipc_ept_cfg {
+	const char *name;           /**< @brief Name of the endpoint */
+	struct ipc_service_cb cb;   /**< @brief Event callback structure */
+	int prio;                   /**< @brief Endpoint priority */
+							    /* If the backend supports priorities*/
+	void *priv;                 /**< @brief Private user data */
+};
```
In this proposal, endpoints are identified only by service name `name`. It is fine and precise enough in two core systems. However, to make this solution generic we would need to extend this identification.

From Linux rpmsg documentation:

> OMAP4, for example, has dual Cortex-A9, dual Cortex-M3 and a C64x+ DSP.

To support such more complex architectures, alongside service name, API user should select which processor is the destination of given IPC endpoint. The simplest solution would be to add one more field in this structure to identify target processor (I'm not sure if there is processor identification abstraction in Zephyr yet). Another one would be to replace `const char *name;` with `struct ipc_ipt_id id;`. This new structure could contain only one field (`name`) for now and could be extended later with target processor identification, when any backend would support multiple target processors.

maje-emb replied:
That' s right, this API proposal is sufficient for two cores, but not for more.
It appears that it will be necessary to add a field to the structure specifying the target processor.  
This information along with the name and priority of the endpoint can be placed in an additional structure. for example:
```
struct ipc_ept_id {
    const char *name;
    int target_core_id;
    int prio;
};

struct ipc_ept_cfg {
    struct ipc_ept_id ept_id;
    struct ipc_service_cb cb;  
    void *priv;
};           
```
Is there a mechanism to identify processors in Zephyr?

hubertmis commented on include/ipc/ipc_service.h:
```diff
@@ -0,0 +1,91 @@
+ *  The same function registers endpoints for both master and slave devices.
+ *
+ *  @param ept Endpoint object.
+ *  @param cfg Endpoint configuration.
```
Please clarify which module is responsible to keep buffers for `name` field.

Is value of `name` copied by this function and caller can free the memory buffer containing this value after this function finishes? Or is the caller responsible to maintain static buffer containing `name` value?

It should be explicitly stated to make sure all backend implementations follow the same rule.

maje-emb replied:
The caller is responsible for allocation and maintenance of memory for `name`. 
The backend implementation only works with a pointer to `name`.
It will be necessary to add information about this.

hubertmis commented on subsys/ipc/rpmsg_multi_instance/Kconfig:
```diff
@@ -0,0 +1,59 @@
+
+# Workaround for not being able to have commas in macro arguments
+DT_CHOSEN_Z_IPC_SHM := zephyr,ipc_shm
+DT_CHOSEN_Z_IPC := zephyr,ipc
+DT_CHOSEN_Z_IPC_TX := zephyr,ipc_tx
+DT_CHOSEN_Z_IPC_RX := zephyr,ipc_rx
```
It seems these values are not used in this implementation
```suggestion
```

maje-emb replied:
You're right. It's not used. IPM channel names are manually filled in.

hubertmis commented on subsys/ipc/rpmsg_multi_instance/Kconfig:
```diff
@@ -0,0 +1,59 @@
+	help
+	  This option specifies size of the memory region to be used
+	  for the OpenAMP IPC shared memory.
+config RPMSG_MULTI_INSTANCE_INIT_PRIORITY
```
Missing newline between entries
```suggestion

config RPMSG_MULTI_INSTANCE_INIT_PRIORITY
```

hubertmis commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,411 @@
+{
+	int status;
+
+	struct rpmsg_mi_ctx *ctx = metal_container_of(vq->vq_dev,
+									  struct rpmsg_mi_ctx,
+									  vdev);
```
nit: Alignment looks strange here

maje-emb replied:
Indeed, for a tab size 8, it looks bad.

hubertmis commented on subsys/ipc/ipc_service_backend/ipc_rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,143 @@
+			endpoints[i].ept.priv   = cfg->priv;
+			endpoints[i].ept.cb     = cfg->cb;
+
+			if (snprintf(endpoints[i].ipm_rx_name, IPM_CH_NAME_LEN, "IPM_%d",
+									MASTER ? 2*i : 2*i+1) < 0) {
+				goto cleanup;
+			}
+			if (snprintf(endpoints[i].ipm_tx_name, IPM_CH_NAME_LEN, "IPM_%d",
+									MASTER ? 2*i+1 : 2*i) < 0) {
+				goto cleanup;
+			}
```
This code looks specific to IPM driver ipm_nrfx_icp. I think it should be generalized by taking IPM instance labels from devicetree or Kconfig.

maje-emb replied:
I think you’re right.

hubertmis commented on subsys/ipc/ipc_service/ipc_service.c:
```diff
@@ -0,0 +1,59 @@
+int ipc_service_register_endpoint(struct ipc_ept **ept, const struct ipc_ept_cfg *cfg)
+{
+	LOG_DBG("Register endpoint %s", cfg->name ? cfg->name : "");
+	if (!backend->register_endpoint) {
```
```suggestion
	if (!backend || !backend->register_endpoint) {
```

hubertmis commented on subsys/ipc/ipc_service/ipc_service.c:
```diff
@@ -0,0 +1,59 @@
+
+int ipc_service_send(struct ipc_ept *ept, const void *data, size_t len)
+{
+	if (!backend->send) {
```
```suggestion
	if (!backend || !backend->send) {
```

hubertmis commented on subsys/ipc/ipc_service/Kconfig:
```diff
@@ -0,0 +1,40 @@
+	bool "IPC service support multiple backends"
+	help
+	  Enables support for a service that ca be shared by multiple
+	  users. Ability to work in different backends. The backend
+	  should be registered before using the IPC Service on the
+	  application side.
```
It looks backend should register itself. It is not registered by the application. I would rephrase this to clarify.

maje-emb replied:
The backend is registered independent of the application. 
All applications that use IPC Service, use the same backend. 
Backend registration is by the `ipc_service_register_backend` function, which is called by `SYS_INIT`.

hubertmis replied:
I see the source of confusion. I understood "The backend should be registered (...) on the application side", but you meant "(...) using the IPC Service on the application side". It's not clear if "on the application side" refers to the registration or service usage.
```suggestion
	  users. Ability to work in different backends. The backend
	  should be registered before application starts using
	  the IPC Service.
```

maje-emb replied:
Exactly. Your suggestion is appropriate.

hubertmis reviewed (commented):
I have reviewed only services during this iteration.
I'll review samples later.

maje-emb commented on include/ipc/ipc_service.h:
```diff
@@ -0,0 +1,91 @@
+ *  @retval -EINVAL when pointer to an endpoint or endpoint configuration is invalid.
+ *  @retval other errno codes depending on the implementation of the backend.
+ */
+int ipc_service_register_endpoint(struct ipc_ept **ept, const struct ipc_ept_cfg *cfg);
```
Yes, in the current implementation the backend manages memory for each endpoint. 
This is for consideration. For sure, the API will be simpler then. 

It seems to me that some functionality has to be in the backend anyway. The current multi-instance RPMsg implementation uses work_q and the stack for that is stored in the backend. 
`K_THREAD_STACK_ARRAY_DEFINE(ipm_stack_area, NUM_ENDPOINTS, WORK_QUEUE_STACK_SIZE);`

I am not sure if `K_THREAD_STACK_DEFINE` can be an endpoint structure field. 
Therefore, it is necessary to set the number of instances in the configuration to allocate memory.
`config RPMSG_MULTI_INSTANCES_NO`
`#define NUM_ENDPOINTS          CONFIG_RPMSG_MULTI_INSTANCES_NO`
One endpoint is assigned to one RPMsg instance.

hubertmis replied:
I see there is `K_THREAD_STACK_MEMBER` for this purpose, however it is deprecated. If it is not possible to move whole memory of given `ipc_ept` object to API user, it is better to keep whole memory inside backend module, like it is implemented now.

maje-emb commented on include/ipc/ipc_service.h:
```diff
@@ -0,0 +1,91 @@
+ *  It is registered during endpoint registration.
+ *  This structure is packed into the endpoint configuration.
+ */
+struct ipc_service_cb {
+	/** @brief Bind was successful
+	 *  @param priv Private user data.
+	 */
+	int (*bound)(void *priv);
+
+	/** @brief New packet arrived.
+	 *  @param data Pointer to data buffer.
+	 *  @param len Length of @a data.
+	 *  @param priv Private user data.
+	 */
+	int (*received)(void *data, size_t len, void *priv);
+
+	/** @brief An error occurred
+	 *  @param message Error message.
+	 *  @param priv Private user data.
+	 */
+	int (*error)(const char *message, void *priv);
+};
```
In this implementation, the callback returns a value. This was done with the idea of passing the value to the backend.
Similar to the callback implementation from RPMsg:
`typedef int (*rpmsg_ept_cb)(struct rpmsg_endpoint *ept, void *data,
			    size_t len, uint32_t src, void *priv);`

It seems to me that it will be necessary to add information to the description what value should be returned.

maje-emb commented on include/ipc/ipc_service.h:
```diff
@@ -0,0 +1,98 @@
+struct ipc_service_cb {
+	/** @brief Bind was successful
+	 *  @param priv Private user data.
+	 *  @retval positive value on success or negative error value on failure
```
You are right, it would be better if callbacks return void. The API will be more flexible to work with different backend implementations.

maje-emb replied:
Fixed.

hubertmis commented on samples/subsys/ipc/ipc_service/remote/prj.conf:
```diff
@@ -0,0 +1,34 @@
+
+# Configuration backend for IPC Service
+CONFIG_IPM=y
+CONFIG_IPM_NRFX=y
```
Like for host project, this file includes multiple entries specific for Nordic SoCs.

maje-emb replied:
Modified: Like for host sample.

hubertmis commented on samples/subsys/ipc/ipc_service/remote/src/main.c:
```diff
@@ -0,0 +1,110 @@
+		},
+	};
+
+	status = ipc_service_register_endpoint(&ept, &ept_cfg);
```
How about creating two endpoints to demonstrate flexibility of IPC Service API?

maje-emb replied:
A second endpoint has been added.

hubertmis commented on samples/subsys/ipc/ipc_service/src/main.c:
```diff
@@ -0,0 +1,118 @@
+			(k_thread_entry_t)app_task,
+			NULL, NULL, NULL, K_PRIO_COOP(7), 0, K_NO_WAIT);
+
+#if defined(CONFIG_SOC_MPS2_AN521) || \
+	defined(CONFIG_SOC_V2M_MUSCA_B1)
+	wakeup_cpu1();
+	k_msleep(500);
+#endif /* #if defined(CONFIG_SOC_MPS2_AN521) */
```
It looks only Nordic SoCs are supported so far. I think this code can be removed.

hubertmis commented on samples/subsys/ipc/rpmsg_multi_instance/prj.conf:
```diff
@@ -0,0 +1,32 @@
+CONFIG_RPMSG_MULTI_INSTANCES_NO=2
+
+CONFIG_IPM=y
+CONFIG_IPM_NRFX=y
```
Similarly to ipc_service sample, this one also includes multiple entries specific to given SoC. Unless this sample is intended to be used only with this SoC, these entries should be moved to SoC-specific configuration extensions.

hubertmis commented on samples/subsys/ipc/rpmsg_multi_instance/remote/prj.conf:
```diff
@@ -0,0 +1,32 @@
+CONFIG_RPMSG_MULTI_INSTANCES_NO=2
+
+CONFIG_IPM=y
+CONFIG_IPM_NRFX=y
```
And here as well we have SoC-specific entries.

maje-emb replied:
Modified

hubertmis commented on samples/subsys/ipc/rpmsg_multi_instance/remote/src/main.c:
```diff
@@ -0,0 +1,200 @@
+struct rpmsg_mi_ept ept_1, ept_2;
+
+
+int boud_cb(void *priv)
+{
+	struct rpmsg_mi_ept *ept = (struct rpmsg_mi_ept *)priv;
+
+	if (ept == &ept_1) {
+		k_sem_give(&bound_ept1_sem);
+	} else if (ept == &ept_2) {
+		k_sem_give(&bound_ept2_sem);
+	}
+	return 0;
+}
+
+int received_cb(void *data, size_t len, void *priv)
+{
+	struct rpmsg_mi_ept *ept = (struct rpmsg_mi_ept *)priv;
+
+	if (ept == &ept_1) {
+		received_data_1 = *((unsigned int *)data);
+		k_sem_give(&data_rx1_sem);
+	} else if (ept == &ept_2) {
+		received_data_2 = *((unsigned int *)data);
+		k_sem_give(&data_rx2_sem);
+	}
+	return 0;
+}
```
I think it would be a little easier to understand if you created two sets of similar functions: `bound1_cb`, `received1_cb`, `bound2_cb`, `received2_cb` without conditional statements inside.

maje-emb replied:
Changes were made.

hubertmis commented on samples/subsys/ipc/rpmsg_multi_instance/remote/src/main.c:
```diff
@@ -0,0 +1,200 @@
+	.ipm_stack_size     = K_THREAD_STACK_SIZEOF(ipm_stack_area_1),
+	.ipm_thread_name    = "ipm_work_q_1",
+	.ipm_work_q_prio    = 0,
+	.ipm_tx_name        = "IPM_0",
+	.ipm_rx_name        = "IPM_1",
```
Looks like Nordic-specific configuration. These values should be retrieved from device tree or Kconfig.

maje-emb replied:
The IPM channel name configuration has been moved to the Kconfig file.

hubertmis commented on samples/subsys/ipc/rpmsg_multi_instance/remote/src/main.c:
```diff
@@ -0,0 +1,200 @@
+
+static struct rpmsg_mi_ept_cfg ept_cfg_2 = {
+	.ept_no = 0,
+	.priv   = &ept_2,
+	.cb     = &cb,
```
```suggestion
	.cb     = &cb,
	.priv   = &ept_2,
```
to keep the same order as in the previous structure

hubertmis commented on samples/subsys/ipc/rpmsg_multi_instance/remote/src/main.c:
```diff
@@ -0,0 +1,200 @@
+	struct rpmsg_mi_ept *ept = (struct rpmsg_mi_ept *)priv;
+
+	if (ept == &ept_1) {
+		received_data_1 = *((unsigned int *)data);
```
It looks you assumed here that endianness for both ends of IPC communication is the same. I'm afraid that's not always true. Could be safer to decode from network to host endianness here and from host to network when sending integer value.

maje-emb replied:
In other IPC samples, this is done similarly. For consistency with other examples, this is how it was done. This can be changed if desired.

hubertmis replied:
I'm fine to keep it as it is. Just found a potential problem during review.

maje-emb replied:
Sure, when working with architectures of different endianness, it is important to keep this in mind.

carlescufi commented:
cc @kyberdin

maje-emb commented on samples/subsys/ipc/ipc_service/prj.conf:
```diff
@@ -0,0 +1,35 @@
+
+# Configuration backend for IPC Service
+CONFIG_IPM=y
+CONFIG_IPM_NRFX=y
```
The specific configuration for the Nordic device has been moved to the .conf file in the board folder.

maje-emb commented on samples/subsys/ipc/ipc_service/prj.conf:
```diff
@@ -0,0 +1,35 @@
+# Configure all TX channels
+CONFIG_IPM_MSG_CH_1_TX=y
+CONFIG_IPM_MSG_CH_3_TX=y
+
```
Modified: Configuration for Nordic SoC in separate file.

hubertmis commented on subsys/ipc/rpmsg_multi_instance/Kconfig:
```diff
@@ -0,0 +1,74 @@
+endchoice
+
+config RPMSG_MULTI_INSTANCES_NO
+	int "Number of RPMSG instances."
```
Should it be defined range now to align with the number of defined instances below?

maje-emb replied:
Right. Range has been added.

hubertmis reviewed (commented):
The last remaining item I would like to address is specifying processor for given endpoint in `ipc_service` API. I described it in details during first review.

It seems all other comments I had are resolved now.

maje-emb commented:
> The last remaining item I would like to address is specifying processor for given endpoint in `ipc_service` API. I described it in details during first review.

I've added a processor target ID in the IPC Service API. I dropped the idea of using additional structures so as not to obscure the interface. 
Apart from that, I added endpoint registrations with the same priority to the same RPMsg instance (this applies to the Rpmsg backend).

peknis commented on include/ipc/rpmsg_multi_instance.h:
```diff
@@ -0,0 +1,162 @@
+
+
+/** @brief Event callback structure
+ *  It is regisered during endpoint registration.
```
```suggestion
 *  It is registered during endpoint registration.
```

maje-emb replied:
Done

peknis commented on include/ipc/rpmsg_multi_instance.h:
```diff
@@ -0,0 +1,162 @@
+	/** @brief New packet arrived
+	 *  @param data Pointer to data buffer
+	 *  @param len Length of @a data
+	 *  @param priv Private user data
```
Just for consistency, add full stop to the end of each parameter description.

maje-emb replied:
Done.

peknis commented on samples/subsys/ipc/rpmsg_multi_instance/README.rst:
```diff
@@ -0,0 +1,71 @@
+.. _Multiple_instance_RPMsg_sample:
+
+Multiple instance of RPMsg sample Application
+################################
```
Title underline too short.

maje-emb replied:
Fixed.

peknis commented on samples/subsys/ipc/ipc_service/README.rst:
```diff
@@ -0,0 +1,70 @@
+.. _IPC_Service_sample:
+
+IPC Service sample Application
+################################
```
Title underline too long.

maje-emb replied:
Fixed.

KAGA164 commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+#define VRING_DESC_SIZEOF(num) ((num) * (sizeof(struct vring_desc)))
+#define VRING_AVAIL_SIZEOF(num) (sizeof(struct vring_avail) + \
+						((num) * sizeof(uint16_t)) + sizeof(uint16_t))
+#define VRING_ALIGN(size, align) (((size) + (align) - 1) & ~((align) - 1))
```
You can just use `ROUND_DOWN` macro

maje-emb replied:
It seems to me that `ROUND_UP` macro should be used.

KAGA164 commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+#define SHM_DEVICE_NAME          "sram0.shm"
+
+#define LOG_MODULE_NAME rpmsg_multi_instance
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_RPMSG_MULTI_INSTANCE_LOG_LEVEL);
```
Maybe just:
```suggestion
LOG_MODULE_REGISTER(rpmsg_multi_instance, CONFIG_RPMSG_MULTI_INSTANCE_LOG_LEVEL);
```

KAGA164 commented on subsys/ipc/ipc_service_backend/ipc_rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,196 @@
+#include <device.h>
+
+#define LOG_MODULE_NAME ipc_rpmsg_multi_instance
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_IPC_SERVICE_LOG_LEVEL);
```
What about:
```suggestion
LOG_MODULE_REGISTER(ipc_rpmsg_multi_instance, CONFIG_IPC_SERVICE_LOG_LEVEL);
```

KAGA164 commented on subsys/ipc/ipc_service/Kconfig:
```diff
@@ -0,0 +1,40 @@
+menuconfig IPC_SERVICE
+	bool "IPC service support multiple backends"
+	help
+	  Enables support for a service that ca be shared by multiple
```
```suggestion
	  Enables support for a service that can be shared by multiple
```

KAGA164 commented on include/ipc/ipc_service.h:
```diff
@@ -0,0 +1,96 @@
+
+/** @brief Endpoint configuration structure. */
+struct ipc_ept_cfg {
+	const char *name;           /**< @brief Name of the endpoint. */
```
I think you shouldn't use a `brief` for structure member. Also you can reconsider comments formatting. For example:
```suggestion
        /** Name of the endpoint. */
	const char *name;
	
	/** Id of the target processor.*/
	int target_core_id; 
	
```

KAGA164 commented on samples/subsys/ipc/ipc_service/remote/src/main.c:
```diff
@@ -0,0 +1,167 @@
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+	int status = 0;
+	unsigned int message = 0U;
```
```suggestion
	uint8_t message = 0U;
```

KAGA164 commented on samples/subsys/ipc/ipc_service/remote/src/main.c:
```diff
@@ -0,0 +1,167 @@
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+	int status = 0;
+	unsigned int message = 0U;
```
```suggestion
	uint8_t message = 0U;
```

KAGA164 commented on samples/subsys/ipc/ipc_service/src/main.c:
```diff
@@ -0,0 +1,169 @@
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+	int status = 0;
+	unsigned int message = 0U;
```
```suggestion
	 uint8_t message = 0U;
```

KAGA164 commented on samples/subsys/ipc/ipc_service/boards/nrf5340dk_nrf5340_cpuapp.conf:
```diff
@@ -0,0 +1,24 @@
+CONFIG_BOARD_ENABLE_CPUNET=y
+
+# Configuration backend for IPC Service
```
```suggestion
# Backend configuration for IPC Service
```

KAGA164 commented on samples/subsys/ipc/ipc_service/remote/src/main.c:
```diff
@@ -0,0 +1,167 @@
+static K_SEM_DEFINE(data_rx1_sem, 0, 1);
+static K_SEM_DEFINE(data_rx2_sem, 0, 1);
+
+static struct ipc_ept *ept_1, *ept_2;
```
I am not sure if our coding standard allows that.
I would declare those variables in separated lines also in other files

maje-emb commented on include/ipc/ipc_service.h:
```diff
@@ -0,0 +1,96 @@
+	const char *name;           /**< @brief Name of the endpoint. */
+	int target_core_id;         /**< @brief Id of the target processor.*/
+	int prio;                   /**< @brief Endpoint priority. */
+				    /* If the backend supports priorities. */
```
The `/* If the backend supports priorities. */` comment refers to the `int prio;` field in the structure. 
A new formatting style has been applied, this should now be clearer.

peknis commented on samples/subsys/ipc/ipc_service/README.rst:
```diff
@@ -0,0 +1,70 @@
+.. _IPC_Service_sample:
+
+IPC Service sample Application
+##############################
```
```suggestion
IPC Service
###########
```
when adding a sample or library, do not mention sample or library in the title of the page (https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#titles)

maje-emb replied:
Thanks. Fixed.

peknis commented on samples/subsys/ipc/ipc_service/README.rst:
```diff
@@ -0,0 +1,70 @@
+Any IPC machanism can by used as a backend. In this sample it is a multiple instance of RPMsg.
+This application demonstrates how to use IPC Service in Zephyr. It is designed
+to demonstrate how to integrate IPC Service with Zephyr both from a build
+perspective and code.
```
General comment: Start each sentence on a new line.

maje-emb replied:
Fixed.

peknis commented on samples/subsys/ipc/ipc_service/README.rst:
```diff
@@ -0,0 +1,70 @@
+********
+
+IPC Service is an abstraction layer. It needs the corresponding backend registered for proper operation.
+Any IPC machanism can by used as a backend. In this sample it is a multiple instance of RPMsg.
```
```suggestion
You can use any IPC mechanism as a backend.
In this sample, it is a multiple instance of RPMsg.
```

peknis commented on samples/subsys/ipc/ipc_service/README.rst:
```diff
@@ -0,0 +1,70 @@
+
+IPC Service is an abstraction layer. It needs the corresponding backend registered for proper operation.
+Any IPC machanism can by used as a backend. In this sample it is a multiple instance of RPMsg.
+This application demonstrates how to use IPC Service in Zephyr. It is designed
+to demonstrate how to integrate IPC Service with Zephyr both from a build
+perspective and code.
```
```suggestion
This sample demonstrates how to integrate IPC Service with Zephyr.
```

peknis commented on samples/subsys/ipc/ipc_service/README.rst:
```diff
@@ -0,0 +1,70 @@
+   :board: nrf5340dk_nrf5340_cpuapp
+   :goals: debug
+
+Open a serial terminal (minicom, putty, etc.) and connect the board with the
```
```suggestion
Open a serial terminal (for example Minicom or PuTTY) and connect the board with the
```
Do not use latin-based abbreviations, such as etc.

peknis commented on samples/subsys/ipc/ipc_service/README.rst:
```diff
@@ -0,0 +1,70 @@
+- Parity: None
+- Stop bits: 1
+
+Reset the board and the following message will appear on the corresponding
+serial port, one is master another is remote:
```
```suggestion
When you reset the development kit, the following messages (one for master and one for remote) will appear on the corresponding serial ports:
```

peknis commented on samples/subsys/ipc/rpmsg_multi_instance/README.rst:
```diff
@@ -0,0 +1,71 @@
+.. _Multiple_instance_RPMsg_sample:
+
+Multiple instance of RPMsg sample Application
+#############################################
```
```suggestion
Multiple instance of RPMsg
##########################
```

peknis commented on samples/subsys/ipc/rpmsg_multi_instance/README.rst:
```diff
@@ -0,0 +1,71 @@
+Overview
+********
+
+Multiple instance of RPMsg is an abstraction created over OpenAMP that makes initialization
+and endpoints creation process easier.
+This application demonstrates how to use multi-instance RPMsg in Zephyr.
```
```suggestion
Multiple instance of RPMsg is an abstraction created over OpenAMP.
It simplifies the  initialization and endpoint creation process.
This sample demonstrates how to use multi-instance RPMsg in Zephyr.
```

peknis commented on samples/subsys/ipc/rpmsg_multi_instance/README.rst:
```diff
@@ -0,0 +1,71 @@
+   :board: nrf5340dk_nrf5340_cpuapp
+   :goals: debug
+
+Open a serial terminal (minicom, putty, etc.) and connect the board with the
```
```suggestion
Open a serial terminal (for example Minicom or PuTTY) and connect the board with the
```

peknis commented on samples/subsys/ipc/rpmsg_multi_instance/README.rst:
```diff
@@ -0,0 +1,71 @@
+- Parity: None
+- Stop bits: 1
+
+Reset the board and the following message will appear on the corresponding
+serial port, one is master another is remote:
```
```suggestion
When you reset the development kit, the following messages (one for master and one for remote) will appear on the corresponding serial ports:
```

KAGA164 reviewed (approved):
Looks good now

maje-emb commented:
@carlescufi @carlocaione Could you please do a review, or request someone to do it.

carlocaione commented on include/ipc/rpmsg_multi_instance.h:
```diff
@@ -0,0 +1,196 @@
+ * @defgroup rpmsg_multiple_instance_api RPMsg multiple instance APIs
+ * @{
+ */
+
+
```
empty line

carlocaione commented on include/ipc/rpmsg_multi_instance.h:
```diff
@@ -0,0 +1,196 @@
+};
+
+/** @brief Struct describing the context of the RPMsg instanece. */
+struct rpmsg_mi_ctx {
```
please use a consistent alignment among structs.

carlocaione commented on include/ipc/rpmsg_multi_instance.h:
```diff
@@ -0,0 +1,196 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_RPMSG_MULTIPLE_INNSTANCE_H_
+#define ZZEPHYR_INCLUDE_RPMSG_MULTIPLE_INNSTANCE_H
```
fix `INNSTANCE` and `ZZEPHYR`

carlocaione commented on include/ipc/rpmsg_multi_instance.h:
```diff
@@ -0,0 +1,196 @@
+#include <metal/sys.h>
+#include <metal/device.h>
+#include <metal/alloc.h>
+
+
```
empty line

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+
+/* Private macros. */
+#define VRING_DESC_SIZEOF(num) ((num) * (sizeof(struct vring_desc)))
+#define VRING_AVAIL_SIZEOF(num) (sizeof(struct vring_avail) + \
+						((num) * sizeof(uint16_t)) + sizeof(uint16_t))
```
shouldn't we take into account the alignment here as well?

maje-emb replied:
It seems that it is not necessary. The alignment is taken into account when calculating the total vring size.

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+#define VRING_FIRST_SUM(num) (VRING_DESC_SIZEOF(num) + VRING_AVAIL_SIZEOF(num))
+
+
+/* Compute size of vring buffer basen on its size and allignment. */
```
`based` and `alignment`

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+
+/* Macro for getting the size of shared memory occupied by single IPC instance. */
+#define SHMEM_INST_SIZE_GET(vring_size) (VDEV_STATUS_SIZE +  \
+					(2 * VIRTQUEUE_SIZE_GET((vring_size))) + \
```
`VRING_COUNT` instead of `2`

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+
+	return (IS_ENABLED(CONFIG_RPMSG_MULTI_INSTANCE_MASTER) ?
+			VIRTIO_CONFIG_STATUS_DRIVER_OK :
+			sys_read8(ctx->shm_status_reg_addr));
```
can we invalidate before reading if the cache management is enabled? Just to account for caches.

maje-emb replied:
Could you please provide more details? Is this related to `CACHE_MANAGEMENT`?

carlocaione replied:
Yes, it is. In some cases your shared memory is behind caches so you need to be sure to properly flush / invalidate it, see https://github.com/OpenAMP/open-amp/pull/293 for more details.

maje-emb replied:
Thanks for the clarification. This has been added.

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+{
+	struct rpmsg_mi_ctx *ctx  = metal_container_of(p_vdev, struct rpmsg_mi_ctx, vdev);
+
+	sys_write8(status, ctx->shm_status_reg_addr);
```
same here. Can we flush if cache is enabled?

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+
+	sys_write8(status, ctx->shm_status_reg_addr);
+}
+
+
```
empty line

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+	struct rpmsg_mi_ctx *ctx = metal_container_of(vq->vq_dev, struct rpmsg_mi_ctx, vdev);
+
+	if (ctx) {
+		status = ipm_send(ctx->ipm_tx_handle, 0, 0, NULL, 0);
```
I strongly suggest to use a user configurable ID here. On some platforms different IDs can have a different meaning on the remote side, so better make this configurable.

maje-emb replied:
Do you mean a SOC identifier such as CONFIG_SOC_NRF5340_CPUAPP or CONFIG_SOC_NRF5340_CPUNET?

carlocaione replied:
No, I mean the ID parameter of the `ipm_send` call. You hardcoded that to `0` but that could not always be available or the correct one to use. In some cases not even all the IDs are available for IPC, so better make that configurable.

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+int rpmsg_mi_configure_shm(struct rpmsg_mi_ctx *ctx,
+					   const struct rpsmg_mi_ctx_cfg *cfg)
+{
+	uint8_t vring_size  = VRING_SIZE_GET(SHM_SIZE);
+	uint32_t shm_addr   = SHMEM_INST_ADDR_AUTOALLOC_GET(SHM_START_ADDR, SHM_SIZE, instance);
+	uint32_t shm_size   = SHMEM_INST_SIZE_AUTOALLOC_GET(SHM_SIZE);
```
please, be consistent with the alignment.

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,453 @@
+	ctx->shm_device.irq_num = 0;
+	ctx->shm_device.irq_info = NULL;
+
+	uint32_t rpmsg_reg_size = 2 * VIRTQUEUE_SIZE_GET(vring_size);
+	uint32_t vring_region_size = VRING_SIZE_COMPUTE(vring_size, VRING_ALIGNMENT);
```
declare the variables at the beginning of the function (I was expecting checkpatch to catch this :)

KAGA164 replied:
Is it a requirement from our coding standard ? I didn't find such requirement here:  https://docs.zephyrproject.org/latest/contribute/index.html#coding-style but I have something like that in my mind.

carlocaione replied:
yeah, probably it's just a "nice-to-have" after all.

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,466 @@
+#define SHM_DEVICE_NAME          "sram0.shm"
+#define IPM_MSG_ID               CONFIG_IPM_MSG_ID
+
+#if CONFIG_CACHE_MANAGEMENT
+#define RPMSG_MI_FLUSH(x)      metal_cache_flush(&x, sizeof(x))
+#define RPMSG_MI_INVALIDATE(x) metal_cache_invalidate(&x, sizeof(x))
+#else
+#define RPMSG_MI_FLUSH(x)      do { } while (0)
+#define RPMSG_MI_INVALIDATE(x) do { } while (0)
+#endif
```
This is not needed, you only have to use `sys_cache_data_range()`

maje-emb replied:
Fixed.

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,460 @@
+#else
+#define VIRTQUEUE_ID 1
+#endif
+
+
```
empty line

carlocaione commented on subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,460 @@
+					   (void *) ctx->shm_device.regions->virt,
+					   ctx->shm_device.regions->size);
+
+		err = rpmsg_init_vdev(&ctx->rvdev,
+				      &ctx->vdev,
+				      ns_bind_cb,
+				      ctx->shm_io,
+				      &ctx->shpool);
+	} else {
+		err = rpmsg_init_vdev(&ctx->rvdev,
+				      &ctx->vdev,
+				      NULL,
+				      ctx->shm_io,
+				      NULL);
```
I have a 100 characters line, why do not use the entire line?

carlocaione commented on subsys/ipc/ipc_service_backend/ipc_rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,195 @@
+
+	struct rpmsg_mi_ept_cfg ept_cfg = {0};
+
+	instances[i].endpoints[e_idx].priv   = cfg->priv;
+	instances[i].endpoints[e_idx].cb     = cfg->cb;
+
+	ept_cfg.cb      = &cb;
+	ept_cfg.priv    = &instances[i].endpoints[e_idx];
+	ept_cfg.name    = cfg->name;
```
alignment

carlocaione commented on subsys/ipc/ipc_service_backend/ipc_rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,195 @@
+		return -EIO;
+	}
+
+	struct rpmsg_mi_ept_cfg ept_cfg = {0};
```
same

carlocaione commented on subsys/ipc/ipc_service_backend/ipc_rpmsg_multi_instance.c:
```diff
@@ -0,0 +1,195 @@
+		instances[i].is_initialized = true;
+	}
+
+	int e_idx = get_available_ept_slot(&instances[i]);
```
move this on top

carlocaione commented on include/ipc/ipc_service.h:
```diff
@@ -0,0 +1,110 @@
+	const char *name;
+
+	/** Id of the target processor. */
+	int target_core_id;
```
In relation to this comment https://github.com/zephyrproject-rtos/zephyr/pull/36120#discussion_r650008407 I'm a bit confused about this.

An endpoint is part of an RPMsg instance and an RPMsg instance can only be used for IPC between two cores. So what is exactly the reason to have this at all?

The problem is that to be able to put two cores in communication we need a shared memory region and each RPMsg instance is only using one shared memory region so it is clear that this API is indeed only able to put two cores in communication using multiple channel / endpoints.

So my follow-up question is: how do we enable real multi-core (>2) communication? Because AFAICT `CONFIG_RPMSG_MULTI_INSTANCE_SHM_BASE_ADDRESS` is actually restricting the range of possible cores to exactly 2.

maje-emb replied:
The `target_core_id` field in the IPC-Service API was added to work with backends that support communication between multiple cores. 

That is right, RPMsg is used for IPC between two cores. Currently the backend implementation using RPMsg does not support working with multiple cores. 

The `target_core_id` field was intended to specify between which cores we are creating the endpoint. That is, the current core and the target core. It is still core-to-core communication, but within the selected two cores. 

It seems to me that in order to implement something like this in the future, separate regions would have to be allocated of shared memory for each pair of cores. The memory would have to be properly described, e.g. in a *.dts file for each core. 

Since currently the `target_core_id` field is not used, it can be removed. In the future when there is a need to create communication channels between multiple cores the IPC Service API will need to be extended.

hubertmis replied:
> An endpoint is part of an RPMsg instance and an RPMsg instance can only be used for IPC between two cores. So what is exactly the reason to have this at all?

I think naming is confusing here. There are IPC Service endpoints, RPMsg instances, and RPMsg endpoints.

I understand it's like you described with RPMsg instances and RPMsg endpoints. There can be one or more RPMsg instances between two cores and each of those instances can include multiple RPMsg endpoints. To enable RPMsg between cores A and B you need separated instance than between cores B and C.

However, IPC Service has only one instance in a single core. This instance can include multiple IPC Service endpoints. Each of these endpoints can be implemented as a separated RPMsg instance (or some of them can share RPMsg instances, but use separated RPMsg endpoints). And this way, IPC Service endpoints from single instance can be used to communicate with different cores (one endpoint to core A, another to core B). To precisely address IPC Service endpoint, you need information about destination core, to select correct RPMsg instance for given IPC Service endpoint.

carlocaione replied:
@maje-emb 

> The `target_core_id` field in the IPC-Service API was added to work with backends that support communication between multiple cores.

Is there any backend able to support multi-core (1 to X)  communication using one single shared memory region?

> It seems to me that in order to implement something like this in the future, separate regions would have to be allocated of shared memory for each pair of cores. The memory would have to be properly described, e.g. in a *.dts file for each core.

More precisely each DTS file for each core should describe a shared memory region for each RPMsg instance (X shared memory regions for X core-to-core IPC).

TBH this is a pretty big limitation IMO. I don't have any particular objection to see this PR merged but I would prefer to see an issue opened already where we can discuss how to extend this service to make it able to support real multi-core IPC.

@hubertmis 

> I think naming is confusing here. There are IPC Service endpoints, RPMsg instances, and RPMsg endpoints.

Yes, sorry. I should have spent more time on the correct naming.

> To enable RPMsg between cores A and B you need separated instance than between cores B and C.

Ok, but AFAICT this is not possible with the current API because currently you can only have one single RPMsg instance per core (aka one single shared memory region), so the B core have to decide if it wants to communicate with A or C exclusively. Is this correct?

hubertmis replied:
> Ok, but AFAICT this is not possible with the current API because currently you can only have one single RPMsg instance per core (aka one single shared memory region), so the B core have to decide if it wants to communicate with A or C exclusively. Is this correct?

I hope this API does not limit backend implementation in such way. I envision potential IPC Service backend implementation which could create n RPMsg instances based on DTS entries for IPC (containing shared memory regions, target CPU, IPM instances) during system initialization. Then IPC Service user does not have to worry about these hardware details. User could simply select `target_cpu` and `endpoint_name`. It would be backend responsibility to map `(target_cpu, endpoint_name)` tuple to given RPMsg instance and endpoint.

What do you think? Could it work?

carlocaione replied:
yes, this is exactly my point. We should extend this API (maybe in a future PR) to get all the needed information from DTS and build all the different RPMsg instances from there. How to do this is up for discussion though, this is why I was suggesting to open an issue where we can discuss all the gory details about this.

maje-emb replied:
> Is there any backend able to support multi-core (1 to X) communication using one single shared memory region?

At this time, the backend does not support 1 to X communication. Currently there is one backed prepared that uses RPMsg. Others may be created in the future.

hubertmis replied:
> So I'll remove `target_core_id` field a the IPC-Service API. Because it's unnecessary.

I agree. We can re-introduce this field later after the proper discussion @carlocaione suggested.

maje-emb replied:
Fixed.
